File: android/src/main/kotlin/com/tradechart/plugin/renderer/VolumeRenderer.kt
package com.tradechart.plugin.renderer

import android.graphics.RectF
import com.tradechart.plugin.engine.ChartFrame
import kotlin.math.max

class VolumeRenderer : ChartLayerRenderer {
    override fun render(frame: ChartFrame) {
        val volumeRect = frame.volumeRect ?: return
        if (!frame.config.showVolume || !frame.viewport.hasVisibleRange()) {
            return
        }

        val bodyWidth = max(1f, frame.viewport.candleBodyWidth() * 0.7f)
        for (index in frame.viewport.visibleStartIndex..frame.viewport.visibleEndIndex) {
            val centerX = frame.viewport.xCenterForIndex(index, frame.mainChartRect.left)
            val volumeTop = frame.viewport.volumeToY(
                frame.candleStore.volumeAt(index),
                volumeRect.top,
                volumeRect.height(),
            )
            val rect = RectF(
                centerX - bodyWidth / 2f,
                volumeTop,
                centerX + bodyWidth / 2f,
                volumeRect.bottom,
            )
            val paint = if (frame.candleStore.closeAt(index) >= frame.candleStore.openAt(index)) {
                frame.theme.volumeBullPaint
            } else {
                frame.theme.volumeBearPaint
            }
            frame.canvas.drawRect(rect, paint)
        }
    }
}

File: ios/Classes/Renderer/VolumeRenderer.swift
import Foundation

final class VolumeRenderer: ChartLayerRenderer {
  func render(frame: ChartFrame) {
    guard frame.config.showVolume, frame.viewport.hasVisibleRange(), let volumeRect = frame.volumeRect else {
      return
    }

    let context = frame.context
    let bodyWidth = max(1, frame.viewport.candleBodyWidth() * 0.7)

    for index in frame.viewport.visibleStartIndex...frame.viewport.visibleEndIndex {
      let centerX = frame.viewport.xCenter(for: index, plotLeft: frame.mainChartRect.minX)
      let volumeTop = frame.viewport.volumeToY(
        frame.candleStore.volume(at: index),
        top: volumeRect.minY,
        height: volumeRect.height
      )
      let rect = CGRect(
        x: centerX - bodyWidth / 2,
        y: volumeTop,
        width: bodyWidth,
        height: volumeRect.maxY - volumeTop
      )
      let isBull = frame.candleStore.close(at: index) >= frame.candleStore.open(at: index)
      context.setFillColor(isBull ? frame.theme.volumeBullColor : frame.theme.volumeBearColor)
      context.fill(rect)
    }
  }
}
